docs/rfcs/021-raii-runtime.md

```diff
@@ -0,0 +1,209 @@
+# Improving runtime by making use of the RAII idiom
+
+## Motivation
+
+Currently, we have several problems regarding the runtime:
+
+- `TenantState` doesn't clearly show what happens with the tenant. Repository and code comments need to describe how state transitions occur and what they mean. It is critical since the code is going to have more states. The same applies to `TimelineState`.
+- Current tenant and timeline access, shutdown design fails to shut down tenants properly, timelines, background tasks, and control their lifetimes, resulting in races.
+- `tasks_mgr` runtime is redundant, and its design allows us to have even more races without noticing.
+
+This RFC describes how to rewrite a runtime in a more RAII fashion, fixing part of its problems.
+
+## Small quality of life changes
```

Comment:
I didn't understand why these primitives are needed. Maybe this is an implementation detail that should be just left out?

Reply:
Probably so. I don't know if I should remove the first bullet point to make the RFC cleaner or still describe the details - they won't hurt. I'd go for the second option.